### PR TITLE
Implement new traits for `<iterator>`

### PR DESCRIPTION
**Pull request by miscco**
This implements the updates to iterator_traits as well as the new readable_traits and incrementable_traits. It further adds the related iter_value_t and iter_difference_t aliases

### CODE REVIEW

**Review by ericniebler — CHANGES_REQUESTED** [review]
minor nits only

File: .upstream-tests/test/support/iterator_traits_cpp17_iterators.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  __host__ __device__ iterator_traits_cpp17_proxy_input_iterator& operator++();
+
+  // this returns legcay_input_iterator, not iterator_traits_cpp17_proxy_input_iterator
+  __host__ __device__ iterator_traits_cpp17_input_iterator operator++(int);
+
+  __host__ __device__ bool operator==(iterator_traits_cpp17_proxy_input_iterator const&) const;
+#if TEST_STD_VER < 20
+  __host__ __device__ bool operator!=(iterator_traits_cpp17_proxy_input_iterator const&) const;
+#endif
+};
```
[the remark this answers is not in the document]
Comment (miscco) [reply .upstream-tests/test/support/iterator_traits_cpp17_iterators.h:55]: Ditto, this is an artifact from me trying to break things up into "small" commits

File: .upstream-tests/test/support/iterator_traits_cpp17_iterators.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  __host__ __device__ iterator_traits_cpp17_iterator operator++(int);
+};
+
+struct iterator_traits_cpp17_proxy_iterator {
+  __host__ __device__ int operator*();
+  __host__ __device__ iterator_traits_cpp17_proxy_iterator& operator++();
+
+  // this returns legcay_iterator, not iterator_traits_cpp17_proxy_iterator
+  __host__ __device__ iterator_traits_cpp17_iterator operator++(int);
+};
```
[the remark this answers is not in the document]
Comment (miscco) [reply .upstream-tests/test/support/iterator_traits_cpp17_iterators.h:25]: That is an artifact of me trying to break everything into small pieces. The iterators get fully upgraded when we actually merge the interator concepts commit, which comes 4 commits later.

File: .upstream-tests/test/support/iterator_traits_cpp17_iterators.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  __host__ __device__ int& operator*();
+  __host__ __device__ iterator_traits_cpp17_iterator& operator++();
+  __host__ __device__ iterator_traits_cpp17_iterator operator++(int);
+};
+
+struct iterator_traits_cpp17_proxy_iterator {
+  __host__ __device__ int operator*();
+  __host__ __device__ iterator_traits_cpp17_proxy_iterator& operator++();
+
+  // this returns legcay_iterator, not iterator_traits_cpp17_proxy_iterator
```
[the remark this answers is not in the document]
Comment (miscco) [reply .upstream-tests/test/support/iterator_traits_cpp17_iterators.h:23]: Ditto

File: .upstream-tests/test/std/iterators/iterator.requirements/iterator.assoc.types/readable.traits/iter_value_t.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+static_assert(check_no_iter_value_t<double>());
+
+struct S {};
+static_assert(check_no_iter_value_t<S>());
+
+struct different_value_element_members {
+  using value_type = int;
+  using element_type = long;
+};
+static_assert(check_no_iter_value_t<different_value_element_members>());
```
[the remark this answers is not in the document]
Comment (ericniebler) [reply .upstream-tests/test/std/iterators/iterator.requirements/iterator.assoc.types/readable.traits/iter_value_t.pass.cpp:67]: *\<bump>*

**Comment by miscco** [comment]
Closing as this has been mirrored to https://github.com/NVIDIA/cccl/pull/158

**Pull request closed without merging**